sound/soc/soc-core.c

```diff
@@ -1320,6 +1320,12 @@ static void soc_remove_link_dais(struct snd_soc_card *card)
 
 	for_each_comp_order(order) {
 		for_each_card_rtds(card, rtd) {
+			struct snd_soc_dai_link *dai_link = rtd->dai_link;
```

Comment:
@plbossart to keep it balanced, should this be done in soc_free_pcm_runtime?

Reply:
there's no good pair, the soc_free_pcm_runtime() should match the soc_new_pcm_runtime().

I really winged this one, but most likely we need a clean-up. @morimoto FYI

Reply:
I looked at this
````c
	/* probe all DAI links on this card */
	ret = soc_probe_link_dais(card);
	if (ret < 0) {
		dev_err(card->dev,
			"ASoC: failed to instantiate card %d\n", ret);
		goto probe_end;
	}

	for_each_card_rtds(card, rtd) {
		ret = soc_init_pcm_runtime(card, rtd);
		if (ret < 0)
			goto probe_end;
	}
````

the dailink .init is called inside of soc_init_pcm_runtime, so I figured I wold release everything in the dual of soc_probe_link_dais, which is soc_remove_links_dais.

Reply:
but we could also add this in snd_soc_pcm_runtime:

````c
void snd_soc_remove_pcm_runtime(struct snd_soc_card *card,
				struct snd_soc_pcm_runtime *rtd)
{
	lockdep_assert_held(&client_mutex);

        struct snd_soc_dai_link *dai_link = rtd->dai_link;

	/* release machine specific resources */
	if (dai_link->exit)
		dai_link->exit(rtd);

	/*
	 * Notify the machine driver for extra destruction
	 */
	if (card->remove_dai_link)
		card->remove_dai_link(card, rtd->dai_link);

	soc_free_pcm_runtime(rtd);
}
EXPORT_SYMBOL_GPL(snd_soc_remove_pcm_runtime);
````
/**

Reply:
yes, remove_pcm_runtime looks like a better place for sure